google/cloud/forseti/scanner/scanners/external_project_access_scanner.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2017 The Forseti Security Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""External project access scanner."""
+# pylint: disable=line-too-long
+import time
+
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.common.gcp_api import api_helpers # noqa=E501
+from google.cloud.forseti.common.gcp_api import api_data_helpers # noqa=E501
+from google.cloud.forseti.common.gcp_api.cloud_resource_manager import CloudResourceManagerClient # noqa=E501
+from google.cloud.forseti.scanner.audit import external_project_access_rules_engine as epa_rules_engine # noqa=E501
+from google.cloud.forseti.scanner.scanners import base_scanner
+from google.cloud.forseti.services.inventory import storage_helpers
+# pylint: enable=line-too-long
+
+LOGGER = logger.get_logger(__name__)
+
+SCOPES = ['https://www.googleapis.com/auth/cloudplatformprojects.readonly']
+
+# Keep a dict of project ancestry where the project ID is the key.
+# This will save the effort of requerying the ancestry of a project
+# if we come across the project ID again.
+DISCOVERED_ANCESTRIES = {}
```

Comment:
It's not good practice to use a mutable global. This can cause problems with multithreaded applications, and is hard to troubleshoot.

Instead, I suggest adding a thread safe memoize wrapper around client.get_project_ancestry() so it either returns the cached value or calls the API if there is a new project_id requested.